bpf/bpf_netdev.c

```diff
@@ -541,7 +556,7 @@ int from_netdev(struct __sk_buff *skb)
 		/* Pass unknown traffic to the stack */
 		return TC_ACT_OK;
 
-#if defined ENABLE_IPSEC && defined ENCAP_IFINDEX
+#ifdef ENABLE_IPSEC
```

Comment:
Which case is this trying to handle?

I presume that in direct routing, we still have a route for each Cilium pod prefix that directs traffic into the cilium_host device.. OK, so a local application attempts to reach to a pod on another node, it arrives to Cilium the first time, it won't hit this because the mark doesn't have the , falls through to the `handle_ipv[46]()` call below which sets the marks and passes up the stack to encrypt. OK great, now it's encrypted and the stack has to figure out where to route the traffic. Either it knows to send the traffic elsewhere directly, or it thinks that the destination is accessible via this interface... at which point we will strip the mark/cb and pass it back up the stack, and then finally the routing will point it the right way?

Quite possibly missing something or thinking of completely the wrong case, in which case I'm interested to learn.

Reply:
In order for it be encrypted it _needs_ to look like a local delivered packet. To do that we mark it and send it up the stack with a specific route matching on the mark to deliver it locally. After its encrypted it will be routed again. We then have a route in the routing table that sends it back to cilium_host clears the marks/etc and back to the stack it goes.

The question is can we remove the second pass through BPF programs by omitting the route... I'll take a look. We might need to clear the mark (otherwise will hit rule again) from ipsec side for this to work but that wasn't supported in 4.9 kernels :(

What I'm not sure is if rule to route table when the route table misses will we go back to lower priority tables. I was thinking that if a table misses then only higher priority tables will be walked but now that you mention it I'm not entirely sure.

Reply:
@joestringer Latest code removes the route and does a bpf based redirect. Going forward we can optimize the redirect out completely and simply call the code directly. But wanted to do that as an optimization.